// mcp-cli/dist/server.js
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema, ToolSchema, } from "@modelcontextprotocol/sdk/types.js";
import { z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import shell from "shelljs";
import path from "path";
import os from "os";
const ToolInputSchema = ToolSchema.shape.inputSchema;
// Get detailed system information
const getDetailedSystemInfo = () => {
    const nodeVersion = process.version;
    const pythonVersion = shell.exec("python --version", { silent: true }).stdout.trim() ||
        shell.exec("python3 --version", { silent: true }).stdout.trim() ||
        "Python not found";
    const osInfo = {
        type: os.type(),
        platform: process.platform,
        release: os.release(),
        version: os.version(),
        arch: process.arch,
        cpus: os.cpus().map(cpu => ({
            model: cpu.model,
            speed: cpu.speed,
        })),
        totalMemory: os.totalmem(),
        freeMemory: os.freemem(),
    };
    const shellInfo = {
        type: shell.exec("echo $SHELL", { silent: true }).stdout.trim(),
        version: shell.exec("$SHELL --version", { silent: true }).stdout.trim(),
    };
    return {
        runtimes: {
            node: {
                version: nodeVersion,
                execPath: process.execPath,
                features: process.features,
            },
            python: {
                version: pythonVersion,
            },
        },
        os: osInfo,
        shell: shellInfo,
        env: process.env,
    };
};
// Shell command input schema with working directory and environment variables
const ExecuteCommandSchema = z.object({
    workingDirectory: z.string().describe("Working directory for command execution"),
    command: z.string().describe("Shell command to execute"),
    args: z.array(z.string()).describe("Command arguments").optional(),
    env: z.record(z.string()).describe("Environment variables to set").optional(),
});
// File operation schemas with working directory
const FileOperationSchema = z.object({
    workingDirectory: z.string().describe("Working directory for file operation"),
    path: z.string().describe("File path"),
});
const CopyFileSchema = z.object({
    workingDirectory: z.string().describe("Working directory for file operation"),
    source: z.string().describe("Source file path"),
    destination: z.string().describe("Destination file path"),
});
const MoveFileSchema = z.object({
    workingDirectory: z.string().describe("Working directory for file operation"),
    source: z.string().describe("Source file path"),
    destination: z.string().describe("Destination file path"),
});
// Directory operation schema with working directory
const DirectoryOperationSchema = z.object({
    workingDirectory: z.string().describe("Working directory for directory operation"),
    path: z.string().describe("Directory path"),
});
// Path validation schema
const PathValidationSchema = z.object({
    targetPath: z.string().describe("Target path to validate"),
    workingDirectory: z.string().describe("Current working directory"),
});
// Available tool names
var ToolName;
(function (ToolName) {
    ToolName["VALIDATE_PATH"] = "validatePath";
    ToolName["EXECUTE_COMMAND"] = "executeCommand";
    ToolName["READ_FILE"] = "readFile";
    ToolName["WRITE_FILE"] = "writeFile";
    ToolName["COPY_FILE"] = "copyFile";
    ToolName["MOVE_FILE"] = "moveFile";
    ToolName["DELETE_FILE"] = "deleteFile";
    ToolName["CREATE_DIRECTORY"] = "createDirectory";
    ToolName["REMOVE_DIRECTORY"] = "removeDirectory";
    ToolName["LIST_DIRECTORY"] = "listDirectory";
    ToolName["GET_SYSTEM_INFO"] = "getSystemInfo";
})(ToolName || (ToolName = {}));
// Error handling function
const handleError = (error) => {
    if (error instanceof Error) {
        return {
            content: [
                {
                    type: "text",
                    text: `Error: ${error.message}`,
                },
            ],
        };
    }
    return {
        content: [
            {
                type: "text",
                text: "An unknown error occurred",
            },
        ],
    };
};
// Result wrapper function
const wrapResult = (success, message, data) => {
    const result = {
        success,
        message,
        ...(data && { data }),
    };
    return {
        content: [
            {
                type: "text",
                text: JSON.stringify(result, null, 2),
            },
        ],
    };
};
// Build environment variables command
const buildEnvCommand = (env = {}) => {
    const envPairs = Object.entries(env).map(([key, value]) => `${key}=${value}`);
    return envPairs.length > 0 ? `npx cross-env ${envPairs.join(" ")}` : "";
};
class PathValidator {
    constructor(allowedPaths) {
        // Normalize all allowed paths to absolute paths
        this.allowedPaths = allowedPaths.map(p => path.resolve(p));
    }
    // Get allowed paths
    getAllowedPaths() {
        return [...this.allowedPaths];
    }
    // Check if a path is within allowed directories
    validatePath(targetPath, workingDirectory) {
        try {
            // Normalize paths
            const absoluteWorkingDir = path.resolve(workingDirectory);
            const absoluteTargetPath = path.resolve(absoluteWorkingDir, targetPath);
            // Check if path is under working directory
            if (!absoluteTargetPath.startsWith(absoluteWorkingDir)) {
                return {
                    isValid: false,
                    absolutePath: absoluteTargetPath,
                    error: `Path must be within working directory: ${absoluteWorkingDir}`,
                };
            }
            // Check if path is under any allowed directory
            const isUnderAllowedPath = this.allowedPaths.some(allowedPath => absoluteTargetPath.startsWith(allowedPath));
            if (!isUnderAllowedPath) {
                return {
                    isValid: false,
                    absolutePath: absoluteTargetPath,
                    error: `Path must be within allowed directories: ${this.allowedPaths.join(", ")}`,
                };
            }
            return {
                isValid: true,
                absolutePath: absoluteTargetPath,
            };
        }
        catch (error) {
            return {
                isValid: false,
                absolutePath: targetPath,
                error: `Invalid path: ${error instanceof Error ? error.message : String(error)}`,
            };
        }
    }
    // Validate multiple paths
    validatePaths(paths, workingDirectory) {
        const results = paths.map(p => {
            const result = this.validatePath(p, workingDirectory);
            return {
                path: p,
                isValid: result.isValid,
                error: result.error,
            };
        });
        return {
            isValid: results.every(r => r.isValid),
            results,
        };
    }
}
export const createServer = (allowedPaths) => {
    const pathValidator = new PathValidator(allowedPaths);
    const server = new Server({
        name: "mcp-shell",
        version: "1.0.0",
    }, {
        capabilities: {
            tools: {},
        },
    });
    // Handle tool list request
    server.setRequestHandler(ListToolsRequestSchema, async () => {
        const tools = [
            {
                name: ToolName.VALIDATE_PATH,
                description: "Validate if a path is within allowed directories",
                inputSchema: zodToJsonSchema(PathValidationSchema),
            },
            {
                name: ToolName.EXECUTE_COMMAND,
                description: "Execute a shell command with environment variables support. All paths in command must be validated first.",
                inputSchema: zodToJsonSchema(ExecuteCommandSchema),
            },
            {
                name: ToolName.READ_FILE,
                description: "Read file content. Path must be validated first.",
                inputSchema: zodToJsonSchema(FileOperationSchema),
            },
            {
                name: ToolName.WRITE_FILE,
                description: "Write content to file. Path must be validated first.",
                inputSchema: zodToJsonSchema(FileOperationSchema.extend({
                    content: z.string().describe("Content to write"),
                })),
            },
            {
                name: ToolName.COPY_FILE,
                description: "Copy a file. Both paths must be validated first.",
                inputSchema: zodToJsonSchema(CopyFileSchema),
            },
            {
                name: ToolName.MOVE_FILE,
                description: "Move a file. Both paths must be validated first.",
                inputSchema: zodToJsonSchema(MoveFileSchema),
            },
            {
                name: ToolName.DELETE_FILE,
                description: "Delete a file. Path must be validated first.",
                inputSchema: zodToJsonSchema(FileOperationSchema),
            },
            {
                name: ToolName.CREATE_DIRECTORY,
                description: "Create a directory. Path must be validated first.",
                inputSchema: zodToJsonSchema(DirectoryOperationSchema),
            },
            {
                name: ToolName.REMOVE_DIRECTORY,
                description: "Remove a directory. Path must be validated first.",
                inputSchema: zodToJsonSchema(DirectoryOperationSchema),
            },
            {
                name: ToolName.LIST_DIRECTORY,
                description: "List directory contents. Path must be validated first.",
                inputSchema: zodToJsonSchema(DirectoryOperationSchema),
            },
            {
                name: ToolName.GET_SYSTEM_INFO,
                description: "Get detailed system information including Node.js, Python versions, OS info, and allowed paths",
                inputSchema: zodToJsonSchema(z.object({})),
            },
        ];
        return { tools };
    });
    // Handle tool call request
    server.setRequestHandler(CallToolRequestSchema, async (request) => {
        const { name, arguments: args } = request.params;
        try {
            if (name === ToolName.VALIDATE_PATH) {
                const validatedArgs = PathValidationSchema.parse(args);
                const result = pathValidator.validatePath(validatedArgs.targetPath, validatedArgs.workingDirectory);
                return wrapResult(result.isValid, result.isValid ?
                    "Path validation successful" :
                    `Path validation failed: ${result.error}`, result);
            }
            if (name === ToolName.EXECUTE_COMMAND) {
                const validatedArgs = ExecuteCommandSchema.parse(args);
                const { workingDirectory, command, args: commandArgs = [], env = {} } = validatedArgs;
                // Validate working directory
                const workingDirValidation = pathValidator.validatePath(workingDirectory, process.cwd());
                if (!workingDirValidation.isValid) {
                    throw new Error(`Working directory validation failed: ${workingDirValidation.error}`);
                }
                // Build command with environment variables using cross-env
                const envCommand = buildEnvCommand(env);
                const fullCommand = `cd "${workingDirValidation.absolutePath}" && ${envCommand} ${command} ${commandArgs.join(" ")}`;
                // Execute command
                const result = shell.exec(fullCommand, { silent: true });
                return wrapResult(result.code === 0, result.code === 0 ? "Command executed successfully" : "Command execution failed", {
                    stdout: result.stdout,
                    stderr: result.stderr,
                    code: result.code,
                    command: fullCommand,
                    env,
                });
            }
            if (name === ToolName.READ_FILE) {
                const validatedArgs = FileOperationSchema.parse(args);
                const pathValidation = pathValidator.validatePath(validatedArgs.path, validatedArgs.workingDirectory);
                if (!pathValidation.isValid) {
                    throw new Error(pathValidation.error);
                }
                if (!shell.test("-f", pathValidation.absolutePath)) {
                    throw new Error("File does not exist");
                }
                const content = shell.cat(pathValidation.absolutePath);
                return wrapResult(true, "File read successfully", { content });
            }
            if (name === ToolName.WRITE_FILE) {
                const validatedArgs = FileOperationSchema.extend({
                    content: z.string(),
                }).parse(args);
                const pathValidation = pathValidator.validatePath(validatedArgs.path, validatedArgs.workingDirectory);
                if (!pathValidation.isValid) {
                    throw new Error(pathValidation.error);
                }
                shell.ShellString(validatedArgs.content).to(pathValidation.absolutePath);
                return wrapResult(true, "File written successfully");
            }
            if (name === ToolName.COPY_FILE) {
                const validatedArgs = CopyFileSchema.parse(args);
                const sourceValidation = pathValidator.validatePath(validatedArgs.source, validatedArgs.workingDirectory);
                const destValidation = pathValidator.validatePath(validatedArgs.destination, validatedArgs.workingDirectory);
                if (!sourceValidation.isValid) {
                    throw new Error(`Source: ${sourceValidation.error}`);
                }
                if (!destValidation.isValid) {
                    throw new Error(`Destination: ${destValidation.error}`);
                }
                if (!shell.test("-f", sourceValidation.absolutePath)) {
                    throw new Error("Source file does not exist");
                }
                const result = shell.cp(sourceValidation.absolutePath, destValidation.absolutePath);
                return wrapResult(result.code === 0, "File copied successfully");
            }
            if (name === ToolName.MOVE_FILE) {
                const validatedArgs = MoveFileSchema.parse(args);
                const sourceValidation = pathValidator.validatePath(validatedArgs.source, validatedArgs.workingDirectory);
                const destValidation = pathValidator.validatePath(validatedArgs.destination, validatedArgs.workingDirectory);
                if (!sourceValidation.isValid) {
                    throw new Error(`Source: ${sourceValidation.error}`);
                }
                if (!destValidation.isValid) {
                    throw new Error(`Destination: ${destValidation.error}`);
                }
                if (!shell.test("-f", sourceValidation.absolutePath)) {
                    throw new Error("Source file does not exist");
                }
                const result = shell.mv(sourceValidation.absolutePath, destValidation.absolutePath);
                return wrapResult(result.code === 0, "File moved successfully");
            }
            if (name === ToolName.DELETE_FILE) {
                const validatedArgs = FileOperationSchema.parse(args);
                const pathValidation = pathValidator.validatePath(validatedArgs.path, validatedArgs.workingDirectory);
                if (!pathValidation.isValid) {
                    throw new Error(pathValidation.error);
                }
                if (!shell.test("-f", pathValidation.absolutePath)) {
                    throw new Error("File does not exist");
                }
                const result = shell.rm(pathValidation.absolutePath);
                return wrapResult(result.code === 0, "File deleted successfully");
            }
            if (name === ToolName.CREATE_DIRECTORY) {
                const validatedArgs = DirectoryOperationSchema.parse(args);
                const pathValidation = pathValidator.validatePath(validatedArgs.path, validatedArgs.workingDirectory);
                if (!pathValidation.isValid) {
                    throw new Error(pathValidation.error);
                }
                const result = shell.mkdir("-p", pathValidation.absolutePath);
                return wrapResult(result.code === 0, "Directory created successfully");
            }
            if (name === ToolName.REMOVE_DIRECTORY) {
                const validatedArgs = DirectoryOperationSchema.parse(args);
                const pathValidation = pathValidator.validatePath(validatedArgs.path, validatedArgs.workingDirectory);
                if (!pathValidation.isValid) {
                    throw new Error(pathValidation.error);
                }
                if (!shell.test("-d", pathValidation.absolutePath)) {
                    throw new Error("Directory does not exist");
                }
                const result = shell.rm("-rf", pathValidation.absolutePath);
                return wrapResult(result.code === 0, "Directory removed successfully");
            }
            if (name === ToolName.LIST_DIRECTORY) {
                const validatedArgs = DirectoryOperationSchema.parse(args);
                const pathValidation = pathValidator.validatePath(validatedArgs.path, validatedArgs.workingDirectory);
                if (!pathValidation.isValid) {
                    throw new Error(pathValidation.error);
                }
                if (!shell.test("-d", pathValidation.absolutePath)) {
                    throw new Error("Directory does not exist");
                }
                const files = shell.ls(pathValidation.absolutePath);
                return wrapResult(true, "Directory listing retrieved successfully", { files });
            }
            if (name === ToolName.GET_SYSTEM_INFO) {
                const systemInfo = getDetailedSystemInfo();
                return wrapResult(true, "System information retrieved successfully", {
                    ...systemInfo,
                    security: {
                        allowedPaths: pathValidator.getAllowedPaths(),
                    },
                });
            }
            throw new Error(`Unknown tool: ${name}`);
        }
        catch (error) {
            return handleError(error);
        }
    });
    return { server };
};
